src/shogun/distance/EuclideanDistance.cpp

```diff
@@ -177,3 +179,69 @@ float64_t CEuclideanDistance::distance_upper_bounded(int32_t idx_a, int32_t idx_
 
 	return result;
 }
+
+SGMatrix<float64_t> CEuclideanDistance::get_distance_matrix()
+{
+	if (lhs->get_feature_class()==C_DENSE && rhs->get_feature_class()==C_DENSE)
+	{
+		if (lhs->get_feature_type()==F_DREAL && rhs->get_feature_type()==F_DREAL)
+		{
+			CDenseFeatures<float64_t>* casted_lhs=static_cast<CDenseFeatures<float64_t>*>(lhs);
+			CDenseFeatures<float64_t>* casted_rhs=static_cast<CDenseFeatures<float64_t>*>(rhs);
+			return compute_distance_matrix(casted_lhs, casted_rhs);
+		}
+
+		else if (lhs->get_feature_type()==F_SHORTREAL && rhs->get_feature_type()==F_SHORTREAL)
+		{
+
+			CDenseFeatures<float32_t>* casted_lhs=static_cast<CDenseFeatures<float32_t>*>(lhs);
+			CDenseFeatures<float32_t>* casted_rhs=static_cast<CDenseFeatures<float32_t>*>(rhs);
+			return compute_distance_matrix(casted_lhs, casted_rhs);
+		}
+
+		else
+			return CDistance::get_distance_matrix();
+	}
+
+	else
+		return CDistance::get_distance_matrix();
+}
+
+template <typename T>
+SGMatrix<float64_t> CEuclideanDistance::compute_distance_matrix(CDenseFeatures<T>* casted_lhs, CDenseFeatures<T>* casted_rhs)
+{
+	int32_t n = casted_lhs->get_num_vectors();
+	int32_t m = casted_rhs->get_num_vectors();
+
+	typename SGMatrix<T>::EigenMatrixXt map_result;
+
+	SGMatrix<T> l_matrix = casted_lhs->get_feature_matrix();
+	SGMatrix<T> r_matrix = casted_rhs->get_feature_matrix();
+	typename SGMatrix<T>::EigenMatrixXtMap map_l(l_matrix.matrix, l_matrix.num_rows, l_matrix.num_cols);
+	typename SGMatrix<T>::EigenMatrixXtMap map_r(r_matrix.matrix, r_matrix.num_rows, r_matrix.num_cols);
+
+	Map<VectorXd> map_lhs_norms(m_lhs_squared_norms.vector, n);
+	typename SGVector<T>::EigenVectorXt casted_lhs_norms = map_lhs_norms.template cast<T>();
+
+	Map<VectorXd> map_rhs_norms(m_rhs_squared_norms.vector, m);
+	typename SGVector<T>::EigenVectorXt casted_rhs_norms = map_rhs_norms.template cast<T>();
+
+	map_result = map_l.transpose()*map_r;
```

Comment:
I think you can have the `map_result` as `double` valued matrix from the beginning (take the matrix that you'll return in the end in the beginning itself, and then use the map as lvalue). So, if the `lhs` and `rhs` vectors are `float32_t` valued, you'll have just one cast (in this above expression). You won't need to cast the norm vectors. 

In the present code, in addition to casting the map result, we're having to cast the vectors (== mem-alloc/copy for those entire vectors) as well. Also, in the end, we're having to copy all the values back.

What do you think?